### DISCUSSION TITLE
Get intensity of collisions

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#191] [Q&A] [answered]
In the physics library we've been previously using (Bullet) a contact manifold also held the applied impulse. Is there a way to get this info from Jolt, too, or should I just calculate it myself from changes in velocity?

**Comment by jrouwe** [comment, marked as answer]
When the collision is detected (and your callback called), the the velocities of the bodies have not been modified yet, so you can get them and combine them with the restitution of the bodies to estimate the applied impulse. The actual impulse depends on constraint solving which happens much later in the pipeline, so they cannot be passed to the callback.

See eq 8-18 in http://www.cs.cmu.edu/afs/cs/user/baraff/www/pbm/rigid2.pdf

I think I have to provide a helper function for this as you're not the first to ask for this (putting this on my TODO list).

**Reply by DocAce** [reply]
In my case I can just as well look at the change in velocity after the simulation step. I only need it for a ballpark estimate of collision damage so precision isn't important. Just wanted to make sure there wasn't some function I was simply not seeing, thanks :)

**Reply by mihe** [reply]
> I think I have to provide a helper function for this as you're not the first to ask for this (putting this on my TODO list).

Did this ever materialize?

I found myself needing a somewhat accurate impulse vector for my implementation and with not really being much of a physics/math guy I only just barely managed to assemble [something that seems to work](https://github.com/godot-jolt/godot-jolt/blob/51386453f71463478ab8f48d9f3340437849ec4d/src/jolt_contact_listener.cpp#L17-L84). All the formulas and stuff in that document you linked is not my strong suit.

A helper function for this in Jolt would be great.

**Reply by jrouwe** [reply]
No it's still on my todo list. I took a look at it:

```
const JPH::MotionProperties& motion1 = *p_body1.GetMotionPropertiesUnchecked();
const JPH::MotionProperties& motion2 = *p_body2.GetMotionPropertiesUnchecked();
````

This is unsafe if one of the bodies is static. A static body will return a nullptr (unless you create all bodies with ```mAllowDynamicOrKinematic = true```).

Other than the function ```calculate_contact_impulse``` looks correct.

The code where you call the function is not correct though, you call it ```mRelativeContactPointsOn1.size()``` times causing an overestimation of the total impulse applied (applying the impulse of the 1st contact would affect velocity of both bodies and thereby influence the results of the 2nd call). I need to think a bit on a proper way to fix this.

Note b.t.w. that impulse A on B = -impulse B on A so you don't need your second loop calculating the impulses as it should be the negative of the values of the first loop.

**Reply by mihe** [reply]
> unless you create all bodies with `mAllowDynamicOrKinematic = true`

Yes, I do just that. Godot has the ability to change motion types on-the-fly, so I'm forced to create bodies with `mAllowDynamicOrKinematic`, which I'm exploiting there.

> Other than the function `calculate_contact_impulse` looks correct.

Glad to hear it!

> impulse A on B = -impulse B on A so you don't need your second loop

Good to know! I wasn't sure if the two arrays "lined up" with one another, but when thinking about it now I guess it wouldn't have made much sense to have two objects collide and end up with two completely different sets of contact points.

**Reply by jrouwe** [reply]
I created a helper function in #435. Note that it doesn't estimate impulses due to friction, but I'm guessing that the godot doesn't do that either as it expects a single float. Friction is perpendicular to the contact impulse so it wouldn't make sense to just add it.

**Reply by mihe** [reply]
> I created a helper function in https://github.com/jrouwe/JoltPhysics/pull/435.

Awesome! From my (limited) testing it seems to work great.

> I'm guessing that the godot doesn't do that either as it expects a single float

Godot 3 (current stable) did indeed just use a float, but Godot 4 does use [a vector](https://github.com/godotengine/godot/blob/0c27edf3d971bd3accffbaee5c57da33d0549284/servers/physics_3d/godot_body_direct_state_3d.cpp#L191-L194), which does [seem to](https://github.com/godotengine/godot/blob/0c27edf3d971bd3accffbaee5c57da33d0549284/servers/physics_3d/godot_body_pair_3d.cpp#L377-L379) incorporate a tangential impulse, which in turn [seems to](https://github.com/godotengine/godot/blob/0c27edf3d971bd3accffbaee5c57da33d0549284/servers/physics_3d/godot_body_pair_3d.cpp#L543-L547) incorporate friction, assuming I'm reading this right. I'm perfectly fine with skipping friction though. I feel like this should already satisfy most of what I would expect people to use this impulse for.

Thank you yet again for the quick turnaround!

**Reply by jrouwe** [reply]
Ah, I see where I went wrong, I looked at [this](https://github.com/godot-jolt/godot-jolt/blob/51386453f71463478ab8f48d9f3340437849ec4d/src/jolt_contact_listener.cpp#L184) line, thinking that you were storing the float in godot's Contact structure, but this is your own class. 

It's not hard to support friction as well, I'll take a look.

**Reply by jrouwe** [reply]
I created #441

**Reply by mihe** [reply]
Thank you. This too seems to be working great. Much appreciated.

Does it make sense for me to try and keep the `inNumIterations` parameter to always be the same value as `PhysicsSettings::mNumVelocitySteps`, if the latter ends up being configurable by the user through some mechanism?

**Reply by jrouwe** [reply]
I would need to check to be sure, but my guess is that anything over 4 is overkill for a decent estimation. 10 is really only needed for proper stacking of many objects but this function only looks at 2 objects.

**Reply by AndreaCatania** [reply]
This helped me as well, thanks a lot!
